app/Platform/Actions/UpdatePlayerPointsStatsBatchAction.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Platform\Actions;
+
+use App\Models\PlayerAchievement;
+use App\Models\PlayerStat;
+use App\Models\System;
+use App\Models\User;
+use App\Platform\Actions\Concerns\CalculatesPlayerPointsStats;
+use Illuminate\Support\Carbon;
+use Illuminate\Support\Facades\DB;
+
+class UpdatePlayerPointsStatsBatchAction
+{
+    use CalculatesPlayerPointsStats;
+
+    /**
+     * Process a batch of users and update their points stats using bulk operations.
+     *
+     * @param array<int> $userIds
+     */
+    public function execute(array $userIds): void
+    {
+        if (empty($userIds)) {
+            return;
+        }
+
+        // Get users and handle untracked users.
+        $users = User::whereIn('id', $userIds)->get()->keyBy('id');
+
+        // Wipe stats for untracked users.
+        $untrackedUserIds = $users->where('Untracked', true)->pluck('id')->toArray();
+        if (!empty($untrackedUserIds)) {
+            PlayerStat::whereIn('user_id', $untrackedUserIds)->delete();
+        }
+
+        // Process tracked users.
+        $trackedUsers = $users->where('Untracked', false);
+        if ($trackedUsers->isEmpty()) {
+            return;
+        }
+
+        $trackedUserIds = $trackedUsers->pluck('ID')->toArray();
+
+        // Fetch all player achievements for tracked users in the time window.
+        $allAchievements = PlayerAchievement::whereIn('player_achievements.user_id', $trackedUserIds)
+            ->whereBetween('player_achievements.unlocked_at', [Carbon::now()->subDays(8), Carbon::now()])
+            ->join('Achievements', 'player_achievements.achievement_id', '=', 'Achievements.ID')
+            ->join('GameData', 'Achievements.GameID', '=', 'GameData.ID')
+            ->where('GameData.ConsoleID', '!=', System::Events)
+            ->select(
+                'player_achievements.user_id',
+                'player_achievements.unlocked_at',
+                'player_achievements.unlocked_hardcore_at',
+                'Achievements.Points as points',
+                'Achievements.TrueRatio as points_weighted'
+            )
+            ->get()
+            ->groupBy('user_id');
```

Comment:
Should this be `orderBy`? You don't have any aggregate columns in your select.

Reply:
I'm pretty sure `groupBy()` is correct here. The key (ambiguous) thing to observe is it's placed after `get()` -- it's not running within the query itself. It's being used as a `Collection` method.

The `groupBy()` is transforming the flat collection of achievements into a nested structure. This is necessary because we later grab the achievements for a given user ID:
```php
$userAchievements = $allAchievements->get($userId, collect());
```

Without the `groupBy()`, we've probably have to spend compute filtering through the entire collection for each user. With `groupBy()`, it's O(1) per each user.